### PR TITLE
Add guidance on Concurrency

### PR DESCRIPTION
**Pull request by cy255**
Adding guidance on concurrency inspired by a WWDC talk on Actors.
Convention added:
Use a task to manage synchronous access to a property populated by an an asynchronous function within an Actor.

### CODE REVIEW

File: Concurrency.md [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+## Rationale
+Even though Actors are helpful for sharing information between concurrent code, asynchronous functions in Actors does not guarantee the state of Actor properties after an asynchronous call.
+
+## Examples
+
+### Bad: when calling the function retrieveURL() across different threads, it's possible that multiple downloads are executed which is inefficient and unnecessary. If the download function always downloads to the same directory, the returned URL might point to an invalid file because another thread has overwritten the url with partially downloaded contents.
+
+``` swift
+actor BuildToolExecutable {
```
Comment (anandrajeswaran) [inline Concurrency.md]: Minor: could genericize this a bit more by not bringing in the complexity that it's an executable. Could it just be a resource that returns Data? Downloading data from a url seems like a slightly more universal and less scary thing than downloading an executable

File: Concurrency.md [hunk 2]
```diff
@@ -0,0 +1,69 @@
+# Concurrency
+
+In general, follow the concurrency section in the [official guidelines](https://docs.swift.org/swift-book/LanguageGuide/Concurrency.html).
+
+## Convention
+
+Use a task to manage synchronous access to a property populated by an an asynchronous function within an Actor. For more in-depth discussions, watch [WWDC21 - Protect mutable state with Swift actors](https://developer.apple.com/wwdc21/10133).
+
+## Rationale
+Even though Actors are helpful for sharing information between concurrent code, asynchronous functions in Actors does not guarantee the state of Actor properties after an asynchronous call.
```
Comment (anandrajeswaran) [inline Concurrency.md]: minor grammar: "asynchronous functions in Actors does not guarantee" -> "asynchronous functions in Actors do not guarantee"

File: Concurrency.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+## Examples
+
+### Bad: when calling the function retrieveURL() across different threads, it's possible that multiple downloads are executed which is inefficient and unnecessary. If the download function always downloads to the same directory, the returned URL might point to an invalid file because another thread has overwritten the url with partially downloaded contents.
+
+``` swift
```
Comment (anandrajeswaran) [inline Concurrency.md:16]: Is there a way in this example we could show the sample code accessing this property and printing it so we could show the order and contents of what is printed to clarify the potential badness?

Comment (cy255) [reply Concurrency.md:16]: Modified the image cache example a little bit to show the potential badness (file contents might be incomplete) in addition to the inefficiency of the "bad" example.

**Review by anandrajeswaran — APPROVED** [review]

File: Concurrency.md [hunk 4]
```diff
@@ -0,0 +1,84 @@
+# Concurrency
+
+In general, follow the concurrency section in the [official guidelines](https://docs.swift.org/swift-book/LanguageGuide/Concurrency.html).
+
+## Convention
+
+Use a task to manage synchronous access to a property populated by an an asynchronous function within an Actor. For more in-depth discussions of the provided example, watch [WWDC21 - Protect mutable state with Swift actors](https://developer.apple.com/wwdc21/10133).
```
Comment (nodes11) [inline Concurrency.md]: nit: "populated by an an asynchronous" --> "populated by an asynchronous"

**Review by nodes11 — APPROVED** [review]

**Pull request merged**